### ISSUE TITLE
memory leaks in parser error handling with strings, all regex and transliterator operators

### ISSUE THREAD

**Issue by davidnich** [qorelanguage/qore#1976]
bug info: 
https://gist.github.com/omusil24/59909393602d8bf9512406f2a6be98b8

**Comment by omusil24** [comment]
related to #1982

**Comment by davidnich** [comment]
@omusil24 this memory leak was introduced with the fix for #1905, therefore I added the `invisible` tag

**Comment by davidnich** [comment]
there is a still a memory leak in all scan expressions where memory is allocated in the start condition and then an EOF happens in that condition - see `examples/test/qore/stack/exception-location.qtest`:

```
==2203== 136 (24 direct, 112 indirect) bytes in 1 blocks are definitely lost in loss record 34 of 45
==2203==    at 0x4C2E1FC: operator new(unsigned long) (vg_replace_malloc.c:334)
==2203==    by 0x50DD628: yylex(YYSTYPE*, QoreParserLocation*, void*) (scanner.lpp:1181)
==2203==    by 0x50C00F7: yyparse(void*) (parser.cpp:6119)
==2203==    by 0x53E22C6: qore_program_private::internParsePending(char const*, char const*, char const*, int) (qore_program_private.h:648)
==2203==    by 0x5342B71: qore_program_private::parse(char const*, char const*, ExceptionSink*, ExceptionSink*, int, char const*, int) (qore_program_private.h:815)
==2203==    by 0x53429CF: qore_program_private::parse(QoreString const*, QoreString const*, ExceptionSink*, ExceptionSink*, int, QoreString const*, int) (qore_program_private.h:799)
==2203==    by 0x519178D: QoreProgram::parse(QoreString const*, QoreString const*, ExceptionSink*, ExceptionSink*, int, QoreString const*, int) (QoreProgram.cpp:979)
...
```

happens in scanner.cpp for example with:
```
\"                                      yylval->String = new QoreStringNode; yylloc->saveFirst(); BEGIN(str_state);
<str_state>{
      \"                                yylloc->restoreFirst(); BEGIN(INITIAL); return QUOTED_WORD;
      \n                                yylval->String->concat('\n');
      {OCTAL_CONST}                     {
                                           int token_length = strlen(yytext);
                                           int val;
                                           if (token_length == 4) {
                                              val = ascii_to_octal(*(yytext + 3))
                                                 + (8 * ascii_to_octal(*(yytext + 2)))
                                                 + (64 * ascii_to_octal(*(yytext + 1)));
                                           }
                                           else if (token_length == 3) {
                                              val = ascii_to_octal(*(yytext + 2))
                                                 + (8 * ascii_to_octal(*(yytext + 1)));
                                           }
                                           else { // here token_length == 2
                                              val = ascii_to_octal(*(yytext + 1));
                                           }
                                           if (val >= 0400)
                                              parse_error(get_loc(yylloc), "trying to escape invalid octal value (in the range 400-777): %o", val);
                                           yylval->String->concat((char)val);
                                        }
      \\n                               yylval->String->concat('\n');
      \\t                               yylval->String->concat('\t');
      \\r                               yylval->String->concat('\r');
      \\b                               yylval->String->concat('\b');
      \\f                               yylval->String->concat('\f');
      \\v                               yylval->String->concat('\v');
      \\.                               yylval->String->concat(yytext[1]);
      \\\n                              yylval->String->concat('\n');
      [^\\\n\"]+                        {
                                           char* yptr = yytext;
                                           while (*yptr)
                                              yylval->String->concat(*(yptr++));
                                        }
}
```

if an EOF happens while processing the string (also regex handling, etc) - example qore code:

```
Program p(PO_NEW_STYLE|PO_REQUIRE_TYPES|PO_STRICT_ARGS);
hash ex = getEx(\p.parse(), "sub t() {\nstring str =\n\"sdfdfssf;\n}", "B");
assertEq(4, ex.line);
assertEq(4, ex.endline);
assertEq("<run-time-loaded: B>", ex.file);
```
